### PR TITLE
RaftUniquenessProvider backed by Hibernate

### PR DESCRIPTION
**Pull request by szymonsztuka**

### CODE REVIEW

File: node/src/main/kotlin/net/corda/node/utilities/AppendOnlyClearablePersistentMap.kt [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+import net.corda.core.utilities.loggerFor
+import java.util.*
+import java.util.concurrent.locks.ReentrantReadWriteLock
+import kotlin.concurrent.read
+import kotlin.concurrent.write
+
+
+/**
+ * Implements an unbound caching layer on top of a table accessed via Hibernate mapping.
+ * The cache can be cleared in bull operation which removed entries in underlying table.
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/utilities/AppendOnlyClearablePersistentMap.kt]: bull -> bulk

Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/utilities/AppendOnlyClearablePersistentMap.kt]: -> which removes the entries...

File: node/src/main/kotlin/net/corda/node/utilities/AppendOnlyClearablePersistentMap.kt [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+ */
+class AppendOnlyClearablePersistentMap<K, V, E, out EK> (
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/utilities/AppendOnlyClearablePersistentMap.kt]: It's not clear to me that this should be all in memory.  Can't you use the existing `AppendOnly...` that does not assume that and add a `clear()`  ?  You can use `allPersisted()` to implement snapshot.

Comment (szymonsztuka) [inline node/src/main/kotlin/net/corda/node/utilities/AppendOnlyClearablePersistentMap.kt]: Yes, I'll change it

File: node/src/main/kotlin/net/corda/node/utilities/AppendOnlyClearablePersistentMap.kt (hunk 1 above)
Comment (szymonsztuka) [inline node/src/main/kotlin/net/corda/node/utilities/AppendOnlyClearablePersistentMap.kt]: Class removed.

File: node/src/main/kotlin/net/corda/node/utilities/PersistentMap.kt [hunk 3]
```diff
@@ -256,4 +256,14 @@ class PersistentMap<K, V, E, out EK> (
         criteriaQuery.select(criteriaQuery.from(persistentEntityClass))
         cache.getAll(session.createQuery(criteriaQuery).resultList.map { e -> fromPersistentEntity(e as E).first }.asIterable())
     }
+
+    override fun clear() {
+        val session = DatabaseTransactionManager.current().session
+        val deleteQuery = session.criteriaBuilder.createCriteriaDelete(persistentEntityClass)
+        deleteQuery.from(persistentEntityClass)
+        synchronized(this) {
```
Comment (szymonsztuka) [inline node/src/main/kotlin/net/corda/node/utilities/PersistentMap.kt]: Maybe is better to drop `synchronized` and add note about not being thread save, as in `AppendOnlyPersistentMap`?

Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/utilities/PersistentMap.kt]: Do we ever call clear ?  Under what circumstance?

**Pull request merged**